src/core/metrics.cc

```diff
@@ -163,205 +182,270 @@ Metrics::EnableGPUMetrics()
+    dcgmerr = dcgmConnect(hostIpAddress, &dcgm_handle_);
+  } else {
+    dcgmerr = dcgmStartEmbedded(DCGM_OPERATION_MODE_MANUAL, &dcgm_handle_);
+  }
+  if (dcgmerr != DCGM_ST_OK) {
+    LOG_WARNING << "error, DCGM unable to start: " << errorString(dcgmerr);
     return false;
   }
 
-  // Create NVML metrics for each GPU
-  for (int didx = 0; didx < dcnt; ++didx) {
-    // Get handle for the GPU
-    cudaDeviceProp gpu_properties;
-    cudaerr = cudaGetDeviceProperties(&gpu_properties, didx);
-    if (cudaerr != cudaSuccess) {
-      LOG_WARNING << "failed to get device properties for device " << didx
-                  << ", GPU metrics will not be available for this device: "
-                  << cudaGetErrorString(cudaerr);
-      continue;
-    }
-
-    char pcibusid_str[64];
-    cudaerr =
-        cudaDeviceGetPCIBusId(pcibusid_str, sizeof(pcibusid_str) - 1, didx);
-    if (cudaerr != cudaSuccess) {
-      LOG_WARNING << "failed to get Bus ID for device " << didx
-                  << ", GPU metrics will not be available for this device: "
-                  << cudaGetErrorString(cudaerr);
-      continue;
+  if (standalone_) {
+    dcgmerr = dcgmUpdateAllFields(dcgm_handle_, 1);
+    if (dcgmerr != DCGM_ST_OK) {
+      LOG_WARNING
+          << "error, DCGM unable to update all fields, GPU metrics will "
+             "not be available: "
+          << errorString(dcgmerr);
+      return false;
     }
+  }
 
-    nvmlDevice_t gpu;
-    nvmlReturn_t nvmlerr = nvmlDeviceGetHandleByPciBusId_v2(pcibusid_str, &gpu);
-    if (nvmlerr != NVML_SUCCESS) {
-      LOG_WARNING << "failed to get device from Bus ID "
-                  << ", GPU metrics will not be available for this device: "
-                  << nvmlErrorString(nvmlerr);
-      continue;
-    }
+  unsigned int all_gpu_ids[DCGM_MAX_NUM_DEVICES];
+  int gpu_count;
+  dcgmerr = dcgmGetAllDevices(dcgm_handle_, all_gpu_ids, &gpu_count);
+  if (dcgmerr != DCGM_ST_OK) {
+    LOG_WARNING << "error, DCGM unable to get device info and count, GPU "
+                   "metrics will not be available: "
+                << errorString(dcgmerr);
+    return false;
+  }
 
-    char gpu_name[NVML_DEVICE_NAME_BUFFER_SIZE + 1];
-    nvmlerr = nvmlDeviceGetName(gpu, gpu_name, NVML_DEVICE_NAME_BUFFER_SIZE);
-    if (nvmlerr == NVML_SUCCESS) {
-      LOG_INFO << "Collecting metrics for GPU " << didx << ": " << gpu_name;
+  // Get CUDA-visible PCI bus ids
+  std::set<std::string> pciBusIds;
+  for (int i = 0; i < gpu_count; ++i) {
+    std::string pciBusId = "0000";  // pad 0's for uniformity
+    char pcibusid_str[64];
+    cudaError_t cudaerr =
+        cudaDeviceGetPCIBusId(pcibusid_str, sizeof(pcibusid_str) - 1, i);
+    if (cudaerr == cudaSuccess) {
+      pciBusId.append(pcibusid_str);
+      pciBusIds.emplace(pciBusId);
     } else {
-      LOG_INFO << "Collecting metrics for GPU " << didx;
+      LOG_WARNING << "error, CUDA unable to get PCI bus information for GPU:"
+                  << i;
```

Comment:
"GPU metrics will not be available for this device" follow same pattern as similar messages